[WIP] Fixing the powernet!  Now with GRAPHS!

WarlockD opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Yes Yes, Powernet sucks.  So giving it an update.  Put in a datum that handles graphs by itself so we can use it for other graph needs (aka pipenet, plumming etc) so that we never get those dangling nodes ever again...hah right

## Why It's Good For The Game
To stop people complaing about powernet

## Changelog
:cl:
refactor: refactored powernet
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

WarlockD committed:
rh lets make it simple

WarlockD committed:
Cleaned up update_icon, cleaned out powernet etc.

actions-user committed:
Automatic changelog compile [ci skip]

bobbahbrown commented on code/modules/power/cable.dm:
```diff
@@ -1,17 +1,21 @@
 //Use this only for things that aren't a subtype of obj/machinery/power
 //For things that are, override "should_have_node()" on them
-GLOBAL_LIST_INIT(wire_node_generating_types, typecacheof(list(/obj/structure/grille)))
+▬GLOBAL_LIST_INIT(wire_node_generating_types, typecacheof(list(/obj/structure/grille)))
```
```suggestion
GLOBAL_LIST_INIT(wire_node_generating_types, typecacheof(list(/obj/structure/grille)))
```

WarlockD replied:
oops.  stray finger hit there

WarlockD committed:
Update code/modules/power/cable.dm

Co-authored-by: Bobbahbrown <bobbahbrown@gmail.com>

WarlockD committed:
Another update

WarlockD committed:
Merge branch 'powernet_graph_addon' of https://github.com/WarlockD/tgstation into powernet_graph_addon

WarlockD committed:
Another push, cleaned up powernet and set up some bitflags for power machines

WarlockD committed:
Ok, getting closer to how I want it to go

Making Powernets the components of the graph and the cables as edges.

WarlockD committed:
workhorse function split_powernet(datum/powernet/PN)

WarlockD committed:
Alrighty, getting ready for compile

Right now need to make the "rebuild evey powernet in the station" function but I think we are ok right now.

WarlockD committed:
Merge branch 'master' into powernet_graph_addon

WarlockD committed:
Ok changing the way graphs are done

WarlockD committed:
More nightmare added

Ok, so changing the way icon is updated.  icon_state will be SEPERATE from the wiring graph.  The wiring graph dosn't care WHERE the wires are, it only cares that they are linked and ajacent.  This way, when shuttles move, you don't have to do a damn thing than just rotate icon_state and run update_icon_state() rather than now, having to recreate the powernet and remap the cables.

LemonInTheDark commented:
Resolves #51059?

WarlockD committed:
Alright graph is setup, a bit of copypasa but I think we got something here

WarlockD committed:
I did it, replaced all cables with laier2 calbes...whhhhy

WarlockD commented:
I hate you all.  That new line merged screwed my branches, I figured out how I am going to do it atleast so watch out for my next pull on this.

LemonInTheDark commented:
holding out for a hero.....
